### DISCUSSION TITLE
Maintained JS bindings ?

### DISCUSSION THREAD

**Discussion by KaKi87** [libui-ng/libui-ng#282] [Q&A] [answered]
Hi,

Is there any currently maintained JS binding library ? Every one mentioned in README is years old.

Thanks

**Comment by joao678** [comment, marked as answer]
I'm making one, here: https://github.com/joao678/libui-js
Every control, event, timer and utility functions already work, the only things to do are proper enums and UiArea crashing + complete implementation

**Reply by KaKi87** [reply]
Running Node 20.17.0 on a Ubuntu 22.04 based distro, I cloned it, installed dependencies, then ran `index.js`, but am experiencing the following error :

```
$ node index.js
/tmp/libui-js/node_modules/koffi/index.js:480
        let lib = native3.load(...args);
                          ^

Error: Failed to load shared library: invalid ELF header
    at obj.load (/tmp/libui-js/node_modules/koffi/index.js:480:27)
    at Object.<anonymous> (/tmp/libui-js/libui/lib.js:13:19)
    at Module._compile (node:internal/modules/cjs/loader:1469:14)
```

Thanks

**Reply by joao678** [reply]
That might be because it's trying to read a .dll file, can you build libui-ng from source, place the compiled .so file inside the libs folder and change a string that contains libui.dll to libui.so?

**Reply by KaKi87** [reply]
Not building stuff from source is one of the main reasons why I'm looking for this, sorry :sweat_smile:

**Reply by joao678** [reply]
fixed, it know includes everything

**Reply by KaKi87** [reply]
Thanks, however it now throws `Segmentation fault (core dumped)` when trying to run `index.js` :sob:

**Reply by joao678** [reply]
I think the table widget has some kind of problem, try again now.

**Reply by KaKi87** [reply]
Amazing, thanks !

**Comment by rubyFeedback** [comment]
Keep us updated if you can every now and then. Other projects may learn from implementations.

**Reply by joao678** [reply]
I've ran into a few limitations with UiArea and now I'm completely rewriting it using Bun ffi instead, Expect updates soon.

**Reply by KaKi87** [reply]
Sounds exciting, I'm looking forward to it ! Will it be backwards-compatible with libui-js ?

**Reply by joao678** [reply]
Probably, the actual code structure could remain unchanged, the only difference is that you'll have to use Bun instead of Node

**Reply by joao678** [reply]
Just pushed a update to libui-js, switched from nodejs to bun, and basic uiArea support (drawing rectangles only for now)

**Comment by joao678** [comment]
Hey, just updated my bindings with support for every function in uiArea + uiDrawText, attributed strings and text layout

**Reply by KaKi87** [reply]
Nice, I'll try it tomorrow !

**Reply by KaKi87** [reply]
Hi,
I take it you're no longer interesting in maintaining this ? :sob:
Thanks

**Reply by joao678** [reply]
Actually I think there isn't anything to do anymore, I believe I implemented everything (as far as I remember)...unless there's something else I missed

**Reply by KaKi87** [reply]
Oh, I got segmentation fault 😅

**Reply by kojix2** [reply]
Where did you get the segmentation?

(Unfortunately, it is difficult to create a perfect libuii-ng binding in a language with garbage collection that does not cause segmentation violations at all. One workaround is to give up the responsibility of freeing the corresponding C libui memory when an object is no longer needed, but that means potential memory leaks.)

**Reply by KaKi87** [reply]
When running `bun run controlsTest.js` on Pop OS 24.04

Are you maintaining a fork too ?

**Reply by kojix2** [reply]
I’m working on Ruby and Crystal bindings, so I don’t know much about JavaScript.
But I have some idea where segmentation faults can happen in libui.

If you are on Linux, you can use gdb/lldb.

Run your program in the debugger, and use the command bt to get a backtrace.
This helps to find where the segmentation fault happens.
You can also post the backtrace result to https://deepwiki.com/libui-ng/libui-ng and ask what is happening or what went wrong.

If you don’t know how to use gdb or lldb, you can ask ChatGPT or search online.
That should help you step by step.
